### PR TITLE
feat(blog): add toc

### PR DESCRIPTION
**Pull request by fiji-flo**
## Summary

Display a ToC on blog posts. 
This also refactors the useStickyHeaderHeight hook.

### Problem

Hard to navigate long blog posts.

### Solution

Reuse the ToC component with some tweaks.

---

### Before

image

### After

image

---

## How did you test this change?

Locally

### CODE REVIEW

File: client/src/blog/post.scss [hunk 1]
```diff
@@ -1,135 +1,128 @@
...
-    display: flex;
-    flex-wrap: wrap;
+.blog-post-container {
+  display: grid;
+  gap: 3rem;
+  grid-template-areas:
+    "post"
+    "newsletter";
+  width: 100%;
+  @media screen and (min-width: $screen-lg) {
```
Comment (LeoMcA) [inline client/src/blog/post.scss:10]: nit: (surprised stylelint doesn't enforce this)

```suggestion

  @media screen and (min-width: $screen-lg) {
```

File: client/src/blog/post.scss [hunk 2]
```diff
@@ -1,135 +1,128 @@
...
+    grid-template-areas:
+      "post toc"
+      "newsletter toc";
+    grid-template-columns: minmax(auto, 100%) minmax(0, 12rem);
   }
-
-  .date-author {
-    column-gap: 1.5rem;
-    padding-left: calc(var(--avatar-size) + var(--author-gap));
+  @media screen and (min-width: $screen-xl) {
```
Comment (LeoMcA) [inline client/src/blog/post.scss:18]: nit
```suggestion

  @media screen and (min-width: $screen-xl) {
```

File: client/src/blog/post.scss [hunk 3]
```diff
@@ -1,135 +1,128 @@
...
+.blog-post-container {
+  display: grid;
+  gap: 3rem;
+  grid-template-areas:
+    "post"
+    "newsletter";
+  width: 100%;
+  @media screen and (min-width: $screen-lg) {
+    display: grid;
+    gap: 3rem;
```
Comment (LeoMcA) [inline client/src/blog/post.scss:12]: these should inherit from above:
```suggestion
```

File: client/src/blog/post.scss [hunk 4]
```diff
@@ -1,135 +1,128 @@
...
+      "newsletter toc";
+    grid-template-columns: minmax(auto, 100%) minmax(0, 12rem);
   }
-
-  .date-author {
-    column-gap: 1.5rem;
-    padding-left: calc(var(--avatar-size) + var(--author-gap));
+  @media screen and (min-width: $screen-xl) {
+    display: grid;
+    gap: 3rem;
```
Comment (LeoMcA) [inline client/src/blog/post.scss:20]: these should also inherit:
```suggestion
```

File: client/src/blog/post.scss [hunk 5]
```diff
@@ -1,135 +1,128 @@
...
+    --offset: var(--top-nav-height);
+    display: none;
 
-    &::after {
-      margin-left: calc(4px - var(--author-gap));
+    .toc-container {
+      position: unset;
+      top: auto;
     }
+    @media screen and (min-width: $screen-lg) {
```
Comment (LeoMcA) [inline client/src/blog/post.scss:35]: nit:
```suggestion
    }

    @media screen and (min-width: $screen-lg) {
```

File: client/src/blog/post.tsx [hunk 6]
```diff
@@ -191,11 +193,18 @@ export function BlogPost(props: HydrationData) {
...
-            lang={doc?.locale}
-          >
+        <main className="blog-post-container container main-page-content">
+          <div className="sidebar-container">
+            <div className="toc-container">
+              <aside className="toc">
+                <nav>
+                  {doc.toc && !!doc.toc.length && <TOC toc={doc.toc} />}
+                </nav>
+              </aside>
```
Comment (LeoMcA) [inline client/src/blog/post.tsx:203]: Do we want all of this to be conditional on `doc.toc && !!doc.toc.length`, or do this aside and nav need to be here permanently for the css to work with the placement properly?

File: client/src/document/hooks.ts [hunk 7]
```diff
@@ -108,35 +108,10 @@ export function useStickyHeaderHeight() {
...
-      return parseFloat(stickyHeaderHeight);
-    } else {
-      console.warn(
-        `[useStickyHeaderHeight] --sticky-header-height has unexpected unit: ${stickyHeaderHeight}`
-      );
-      return 0;
-    }
+    return (
+      document.querySelector<HTMLElement>(".main-document-header-container")
+        ?.offsetHeight || 0
```
Comment (LeoMcA) [inline client/src/document/hooks.ts:113]: We might need a fudge factor here of 1 or 2px, as currently clicking on an element in the TOC doesn't highlight it, but scrolling down marginally more does.

File: client/src/document/organisms/toc/index.tsx [hunk 8]
```diff
@@ -10,8 +10,9 @@ export function TOC({ toc }: { toc: Toc[] }) {
   const observedElements = React.useCallback(() => {
     const mainElement = document.querySelector("main") ?? document;
     const elements = mainElement.querySelectorAll(
-      "h1, h1 ~ *:not(section), h2, h2 ~ *:not(section), h3, h3 ~ *:not(section)"
+      "h1, h1 ~ *:not(section), h2:not(.document-toc-heading), h2:not(.document-toc-heading) ~ *:not(section), h3, h3 ~ *:not(section)"
     );
+    console.log(elements);
```
Comment (LeoMcA) [inline client/src/document/organisms/toc/index.tsx:15]: :eyes:

**Review by LeoMcA — APPROVED** [review]
mostly looks good from my slightly time pressured review, mostly nits, give it a good test before deploying to prod!

File: client/src/blog/post.tsx (hunk 6 above)
Comment (fiji-flo) [reply client/src/blog/post.tsx:203]: That's a copy of the article toc. And I'm already touching to many things here, sorry.

**Review by caugner — CHANGES_REQUESTED** [review]
Overall LGTM, but I have some naming suggestions.

However, I noticed when trying this out locally the TOC entries flash blue when clicked (but only in the blog):
image

**Review by caugner — CHANGES_REQUESTED** [review]
⚠️ Also just noticed that when navigating to a section via a TOC entry, the previous TOC entry is getting highlighted after the navigation has finished:
image

**Review by caugner — CHANGES_REQUESTED** [review]
LGTM, but I'm still experiencing the mentioned issues:

https://github.com/mdn/yari/assets/495429/498f570d-2490-4962-aae6-98cdb127d9d4

**Comment by fiji-flo** [comment]
Sorry I fixed that now.

**Review by caugner — APPROVED** [review]
Pushed a fix (73d492f8a92e7801ac0a38d59c3620245fdf890f, the TOC links were not readable while clicking on them, due to `.main-page-content` style being applied to them), now it looks good.

**Pull request merged**